actions/get-vault-secrets/action.yaml

```diff
@@ -0,0 +1,46 @@
+name: Get Vault Secrets
+description: Composite action (step) to get secrets from Grafana Labs' Vault instance.
+inputs:
+  workload_identity_provider:
+    required: true
+    description: Set to `secrets.WORKLOAD_IDENTITY_POOL_PROVIDER`
+  iap_service_account:
+    required: true
+    description: Set to `secrets.VAULT_IAP_SA_EMAIL`
+  iap_audience:
+    required: true
+    description: Set to `secrets.VAULT_IAP_OAUTH_CLIENT_ID`
+  vault_url:
+    required: true
+    description: Set to `secrets.VAULT_URL`
+  secrets:
+    required: true
+    description: Secret mapping. See https://github.com/hashicorp/vault-action for the format.
+
+runs: 
+  using: composite
+  steps:
+    - id: vault-iap-auth
+      uses: google-github-actions/auth@v1
+      with:
+        workload_identity_provider: "${{ inputs.workload_identity_provider }}"
+        service_account: "${{ inputs.iap_service_account }}"
+        token_format: id_token
+        id_token_audience: "${{ inputs.iap_audience }}"
+        id_token_include_email: true
+
+      # Get the secrets
+    - name: Import Secrets
+      id: import-secrets
+      uses: hashicorp/vault-action@v2
+      with:
+        url: "${{ inputs.vault_url }}"
+        role: vault-github-actions
+        path: github-actions-oidc
+        method: jwt
+        jwtGithubAudience: "${{ inputs.vault_url }}"
```

Comment:
tl;dr audience means "who is this token for?" and customising it is a way of scoping down tokens.

The flow

 1. We ask GitHub to give us a token to be handed to the Vault instance. That token is used on Vault's side to check we are allowed to do the action.
 2. Vault checks the issuer and the signature to make sure the token comes from GitHub.
 3. Vault checks the expiry (`exp`) claim to make sure it's still in date.
 4. Vault then checks other claims according to the policy - the interesting one for this comment is the audience. Since we've set it to the URL we can check _was this token meant to be used to authenticate with this Vault?_

Why care?

1. It helps to prevent accidents: if you issue a token and send it to Vault by mistake, it won't work unless you set this audience.
5. If someone manages to get their hands on a token (say we leak them in logs by mistake) for a different workflow, they won't be able to use it to talk to Vault because the audience will be wrong. Check [the claims that are set on the token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token).  If we didn't set the audience we wouldn't be able to tell one token from another on the receiving side. Said another way: it helps to limit the scope of any compromise if we are diligent about setting the audience. The same would apply the other way around: a Vault token wouldn't work anywhere else since the audience is set to Vault.

For workload identity federation we do similar (actually it happens by default): there the audience is set to the URL of the pool.

In the Vault policy itself, we've got it configured to look at the `repository` claim on the token to decide which repo-specific secrets a request can access. So if you get your hands on a _Vault_ access token, the potential compromise is still only limited to that repo.

They've got a good point on that DO post about logs though - that would be interesting to look into. I'll note it on the issue. 👍